napari/_qt/widgets/qt_viewer_buttons.py

```diff
@@ -195,6 +210,209 @@ def open_perspective_popup(self):
+        stride_help_txt = f"<FONT> {stride_help_msg}</FONT>"
+
+        shape_help_symbol.setObjectName("help_label")
+        shape_help_symbol.setToolTip(shape_help_txt)
+
+        stride_help_symbol.setObjectName("help_label")
+        stride_help_symbol.setToolTip(stride_help_txt)
+
+        # layout
+        form_layout = QFormLayout()
+
+        form_layout.insertRow(
+            0,
+            QLabel(trans._('Grid stride:'), parent=popup),
+            grid_stride,
+        )
+
+        form_layout.insertRow(
+            1,
+            QLabel(trans._('Grid width:'), parent=popup),
+            grid_width,
+        )
+
+        form_layout.insertRow(
+            2,
+            QLabel(trans._('Grid height:'), parent=popup),
+            grid_height,
+        )
+
+        help_layout = QVBoxLayout()
+        help_layout.addWidget(stride_help_symbol)
+        help_layout.addWidget(blank)
+        help_layout.addWidget(shape_help_symbol)
+
+        layout = QHBoxLayout()
+        layout.addLayout(form_layout)
+        layout.addLayout(help_layout)
+
+        popup.frame.setLayout(layout)
+
+        popup.show_above_mouse()
+
+        # adjust placement of shape help symbol.  Must be done last
+        # in order for this movement to happen.
+        delta_x = 0
+        delta_y = -15
+        shape_pos = (
+            shape_help_symbol.x() + delta_x,
+            shape_help_symbol.y() + delta_y,
+        )
+        shape_help_symbol.move(QPoint(*shape_pos))
+
+    def _update_grid_width(self, value):
+        """Update the width value in grid shape.
+
+        Parameter
+        ---------
+        value: int
+            New grid width value.
+        """
+
+        # need to skip 0
+        if value == 0:
+            value = self._skip_zero(value, self.last_width_value)
+            self.grid_width_box.setValue(value)
+
+        self.last_width_value = value
+        self.viewer.grid.shape = (self.viewer.grid.shape[0], value)
+
+    def _update_grid_stride(self, value):
+        """Update stride in grid settings.
+
+        Parameter
+        ---------
+        value: int
+            New grid stride value.
+        """
+
+        # need to skip 0
+        if value == 0:
+            value = self._skip_zero(value, self.last_stride_value)
+            self.grid_stride_box.setValue(value)
+
+        self.last_stride_value = value
+
+        self.viewer.grid.stride = value
+
+    def _update_grid_height(self, value):
+        """Update height value in grid shape.
+
+        Parameter
+        ---------
+        value: int
+            New grid height value.
+        """
+
+        # need to skip 0
+        if value == 0:
+            value = self._skip_zero(value, self.last_height_value)
+            self.grid_height_box.setValue(value)
+
+        self.last_height_value = value
+        self.viewer.grid.shape = (value, self.viewer.grid.shape[1])
+
+    def _skip_zero(self, value, last_value):
```

Comment:
QSpinbox already provides enough methods to do this natively without having to store the last value and add methods and stuff.  use a combination of [`stepBy`](https://doc.qt.io/qt-5/qabstractspinbox.html#stepBy) and [`validate`](https://doc.qt.io/qt-5/qabstractspinbox.html#validate):

```py
class SB(QSpinBox):
    def validate(self, input: str, pos: int):
        if input == '0':
            return QValidator.Invalid, input, pos
        return super().validate(input, pos)

    def stepBy(self, steps: int) -> None:
        if (self.value() == 1 and steps == -1) or (
            self.value() == -1 and steps == 1
        ):
            steps *= 2
        return super().stepBy(steps)
```

Reply:
Got this working!   Thanks for this!